govcd/adminvdc.go

```diff
@@ -52,33 +95,241 @@ func (adminVdc *AdminVdc) Refresh() error {
+	if err != nil {
+		return Task{}, err
+	}
+	producer, ok := vdcProducerByVcdVersion["vdc"+apiVersionToVcdVersion[apiVersion]]
+	if !ok {
+		return Task{}, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if producer.UpdateVdcAsync == nil {
+		return Task{}, fmt.Errorf("function UpdateVdcAsync is not defined for %s", "vdc"+apiVersion)
+	}
+	util.Logger.Printf("[DEBUG] UpdateAsync call function for version %s", producer.SupportedVersion)
 
-	adminVdc.AdminVdc.Xmlns = types.XMLNamespaceVCloud
+	return producer.UpdateVdcAsync(adminVdc)
 
-	// Return the task
-	return adminVdc.client.ExecuteTaskRequest(adminVdc.AdminVdc.HREF, http.MethodPut,
-		types.MimeAdminVDC, "error updating VDC: %s", adminVdc.AdminVdc)
 }
 
 // Update function updates an Admin VDC from current VDC struct contents.
 // Any differences that may be legally applied will be updated.
 // Returns an empty AdminVdc struct and error if the call to vCD fails.
 // API Documentation: https://vdc-repo.vmware.com/vmwb-repository/dcr-public/7a028e78-bd37-4a6a-8298-9c26c7eeb9aa/09142237-dd46-4dee-8326-e07212fb63a8/doc/doc/operations/PUT-Vdc.html
 func (adminVdc *AdminVdc) Update() (AdminVdc, error) {
-	task, err := adminVdc.UpdateAsync()
+	apiVersion, err := adminVdc.client.maxSupportedVersion()
 	if err != nil {
 		return AdminVdc{}, err
 	}
 
-	err = task.WaitTaskCompletion()
+	producer, ok := vdcProducerByVcdVersion["vdc"+apiVersionToVcdVersion[apiVersion]]
+	if !ok {
+		return AdminVdc{}, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if producer.UpdateVdc == nil {
+		return AdminVdc{}, fmt.Errorf("function UpdateVdc is not defined for %s", "vdc"+apiVersion)
+	}
+
+	util.Logger.Printf("[DEBUG] Update call function for version %s", producer.SupportedVersion)
+
+	updatedAdminVdc, err := producer.UpdateVdc(adminVdc)
 	if err != nil {
 		return AdminVdc{}, err
 	}
+	return *updatedAdminVdc, err
+}
+
+// CreateOrgVdc creates a VDC with the given params under the given organization
+// and waits for the asynchronous task to complete.
+// Returns an AdminVdc pointer and an error.
+func (adminOrg *AdminOrg) CreateOrgVdc(vdcConfiguration *types.VdcConfiguration) (*Vdc, error) {
+	apiVersion, err := adminOrg.client.maxSupportedVersion()
+	if err != nil {
+		return nil, err
+	}
+	producer, ok := vdcProducerByVcdVersion["vdc"+apiVersionToVcdVersion[apiVersion]]
+	if !ok {
+		return nil, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if producer.CreateVdc == nil {
+		return nil, fmt.Errorf("function CreateVdc is not defined for %s", "vdc"+apiVersion)
+	}
+
+	util.Logger.Printf("[DEBUG] CreateOrgVdc call function for version %s", producer.SupportedVersion)
+	return producer.CreateVdc(adminOrg, vdcConfiguration)
+}
+
+// CreateOrgVdcAsync creates a VDC with the given params under the given organization.
+// Returns a Task and an error.
+func (adminOrg *AdminOrg) CreateOrgVdcAsync(vdcConfiguration *types.VdcConfiguration) (Task, error) {
```

Comment:
Correct me if I'm wrong, but previously CRUD functions for VDC were in `adminorg.go`. With these changes we have functions for older vCD versions in `adminorg.go`, but functions for vCD 9.7+ in `adminvdc.go`. Can we have the same type of functions in a single file, or is there a good reason behind such move?

CC: @dataclouder @Didainius

Reply:
In early days we try to keep with hierarchy type of storing functions in parent if parent is `receiver`, but later we chose directly hold functions in files according type. So everything we do new goes to their own file, not in parents. Version 3.0 is good option to cleanup and move functions accordingly.

Reply:
Is there a reason not to move the touched functions from `adminorg.go` to `adminvdc.go` those functions now, so we have them in a single place and have a tidiness from the perspective of CRUD functions? Now it's pretty hard to navigate the code trying to understand where are 9.7+ functions and where are the older ones.

Reply:
I can move

Reply:
done

Reply:
> In early days we try to keep with hierarchy type of storing functions in parent if parent is receiver, but later we chose directly hold functions in files according type. So everything we do new goes to their own file, not in parents.

> I can move

> done

I was actually referring only to the `CreateVdc` and `CreateVdc` functions, so they'd all be in a single file with other `Create...` ones, but I'm OK with this move if everyone else is. Summary:

131 lines moved from `adminvdc.go` to `adminorg.go`, functions that moved:
- GetAdminVdcByName
- GetAdminVDCByHref
- GetAdminVDCByName
- GetAdminVDCById
- GetAdminVDCByNameOrId
- CreateVdc
- CreateVdc

Commit: https://github.com/vmware/go-vcloud-director/commit/65965d28eb114e83df53d1fd9fa9dfb2ca8c9922

CC: @Didainius @dataclouder